src/compressor/QatAccel.cc

```diff
@@ -48,22 +53,112 @@ bool QatAccel::init(const std::string &alg) {
   rc = qzSetDefaults(&params);
   if (rc != QZ_OK)
       return false;
+  return true;
+}
 
+static bool setup_session(QatAccel::session_ptr &session, QzSessionParams_T &params) {
+  int rc;
   rc = qzInit(session.get(), QZ_SW_BACKUP_DEFAULT);
-  if (rc != QZ_OK && rc != QZ_DUPLICATE && rc != QZ_NO_HW)
+  if (rc != QZ_OK && rc != QZ_DUPLICATE)
     return false;
-
   rc = qzSetupSession(session.get(), &params);
-  if (rc != QZ_OK && rc != QZ_DUPLICATE && rc != QZ_NO_HW ) {
-    qzTeardownSession(session.get());
-    qzClose(session.get());
+  if (rc != QZ_OK) {
     return false;
   }
+  return true;
+}
+
+// put the session back to the session pool in a RAII manner
+struct cached_session_t {
+  cached_session_t(QatAccel* accel)
+    : accel{accel} {
+    std::unique_lock lock{accel->mutex};
+    if (!accel->sessions.empty()) {
+      session = std::move(accel->sessions.back());
+      accel->sessions.pop_back();
+      return;
+    }
 
+    // If there are no available session to use, we try allocate a new
+    // session. When we failed to allocate a new session, we block on here,
+    // waiting a available session;
+    QzSessionParams_T params = {(QzHuffmanHdr_T)0,};
+    QatAccel::session_ptr s{new struct QzSession_S()};
+    memset(s.get(), 0, sizeof(struct QzSession_S));
+    if (get_qz_params(accel->alg_name, params) && setup_session(s, params)) {
+      session = std::move(s);
```

Comment:
i tend to think that errors here should be returned directly to `QatAccel::compress()` or `decompress()`, instead of falling back to `accel->sessions`. that would remove the need for a condition variable entirely

Reply:
I removed a condition variable entirel, so there directly acquire a new session.